---
title: "Biol 432 Group project-Question 1"
author: "Chenyang Wu"
date: "2022/3/21"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Project Info
#### **Group name**: Teambits
#### **Date**: 2022/3/21
#### **GitHub Link**: https://github.com/carolinetang77/BIOL432-group1


## **Research question**: Are mutation rates correlated with transmission method? Do certain transmission methods have higher mutation rates?

### Load the packages we will need
```{r message=FALSE, warning=FALSE}
library(dplyr) 
library(ggplot2)
library(BiocManager)
library(genbankr)
library(rentrez)
library(muscle)
library(ape)
library(reshape2)
library(ggtree)
```

### Input the data
```{r}
Table1 <- read.csv("./InputData/TableS1.csv")
Table9 <- read.csv("./InputData/TableS9.csv")

readLines("./InputData/TableS8.dat", n = 10) # Take a look at the .dat file
Table8 <- read.table("./InputData/TableS8_edited.dat", header = F) 

Table11 <- read.csv("./InputData/TableS11.csv")
```

```{r}
# Check how many types of virus in table S9
table(Table9$virus)

# Check the corresponding transmission type to the abbreviation in table S1
list(Table1$Abbreviation)
list(Table1$Transmission)
```

### Combine the relevant columns into 1 dataframe
```{r}
# Label the transmission type to the dataset
dNdSData <- Table9 %>% 
  mutate(Transmission_Type = recode(virus, 
                                    "BCoV1" = "aerosolic", 
                                    "CHIKV" = "vector", 
                                    "DENV" = "vector", 
                                    "Ebola" = "body_fluids", 
                                    "EVA" = "fecal-oral", 
                                    "EVB" = "fecal-oral", 
                                    "EVC" = "fecal-oral",
                                    "EVCr" = "fecal-oral",
                                    "EVD" = "fecal-oral", 
                                    "H3N2" = "aerosolic", 
                                    "HCV" = "blood/sexual",
                                    "HCVr" = "blood/sexual",
                                    "HDV" = "blood/sexual",
                                    "HMPV" = "aerosolic", 
                                    "HRSV" = "aerosolic", 
                                    "HRV3" = "aerosolic", 
                                    "MERS" = "aerosolic",
                                    "MMV" = "aerosolic", 
                                    "MRV" = "aerosolic",
                                    "Norwalk" = "fecal-oral",
                                    "OHVA" = "fecal-oral", 
                                    "PeVA" = "aerosolic",
                                    "RVA" = "aerosolic",
                                    "SARS2" = "aerosolic",   
                                    "SV" = "fecal-oral",
                                    "TBEV" = "vector",
                                    "WNV" = "vector", 
                                    "YFV" = "vector",
                                    "ZIKV" = "vector")) 

# Draw a box plot for mean dNdS data with different transmission types
ggplot(dNdSData, aes(x = Transmission_Type, y = meandNdS, 
                     na.rm = TRUE, fill = Transmission_Type)) + 
  geom_boxplot(alpha = 0.8) +
  theme_bw() + 
  scale_fill_brewer(palette = "Set2") + 
  labs(x = "Transmission Types", y = "Mean dNdS")
```

#### **Fig.1** Boxplot of mean dNdS values among different transmission types.


```{r}
MutaRate <- Table8 %>% 
  mutate(Transmission_Type = recode(V1, 
                                    "BCoV1" = "aerosolic", 
                                    "CHIKV" = "vector", 
                                    "DENV" = "vector", 
                                    "Ebola" = "body_fluids", 
                                    "EVA" = "fecal-oral", 
                                    "EVB" = "fecal-oral", 
                                    "EVC" = "fecal-oral",
                                    "EVCr" = "fecal-oral",
                                    "EVD" = "fecal-oral", 
                                    "H3N2" = "aerosolic", 
                                    "HCV" = "blood/sexual",
                                    "HCVr" = "blood/sexual",
                                    "HDV" = "blood/sexual",
                                    "HMPV" = "aerosolic", 
                                    "HRSV" = "aerosolic", 
                                    "HRV3" = "aerosolic", 
                                    "MERS" = "aerosolic",
                                    "MMV" = "aerosolic", 
                                    "MRV" = "aerosolic",
                                    "Norwalk" = "fecal-oral",
                                    "OHVA" = "fecal-oral", 
                                    "PeVA" = "aerosolic",
                                    "RVA" = "aerosolic",
                                    "SARS2" = "aerosolic",   
                                    "SV" = "fecal-oral",
                                    "TBEV" = "vector",
                                    "WNV" = "vector", 
                                    "YFV" = "vector",
                                    "ZIKV" = "vector")) 

# Draw a box plot for the mutation rate under different transmission types
ggplot(MutaRate, aes(x = Transmission_Type, 
                     y = V2, na.rm = TRUE, 
                     fill = Transmission_Type)) + 
  geom_boxplot(alpha = 0.8) +
  theme_bw() + 
  scale_fill_brewer(palette = "Set2") + 
  labs(x = "Transmission Types", y = "Mutation Rate")
```

#### **Fig.2** Boxplot of mutation rate among different Transmission Types 


```{r}
# Check the number of IDs in table S11 for each virus
table(Table11$virus)
# Since there are so many IDs, We decided to pick 5 IDs from each virus

# Random pick 5 IDs for each virus
set.seed(1) # Set seed to make sure the output are constant
IDs <- Table11 %>%
  group_by(virus) %>%
  sample_n(5)
```

```{r}
# Double check if the code correctly pick 10 IDs from each virus
table(IDs$virus)

# Create the id list
ncbi_ids <- IDs$ID 
```


```{r}
# Search the sequence info from NCBI
Q1Vir <- entrez_fetch(db = "nuccore", id = ncbi_ids, rettype = "fasta") 
Q1Seq <- strsplit(Q1Vir, split = "\n\n", fixed = T)
Q1Seq <- unlist(Q1Seq)

# Use regular expression to edit the search result
header <- gsub("(^>.*genome|*cds|*sequence|*SEQUENCES|*RNA)\\n[ATCG].*", "\\1", Q1Seq)

seq <- gsub("^>.*genome\\n([ATCG].*)", "\\1", Q1Seq)
seq <- gsub("^>.*cds\\n([ATCG].*)", "\\1", seq)
seq <- gsub("^>.*sequence\\n([ATCG].*)", "\\1", seq)
seq <- gsub("^>.*SEQUENCES\\n([ATCG].*)", "\\1", seq)
seq <- gsub("^>.*RNA\\n([ATCG].*)", "\\1", seq)

Q1SeqTable <- data.frame(Name = header, Sequence = seq)
Q1SeqTable$Sequence <- gsub("\n", "", Q1SeqTable$Sequence)

# There are several lines have different ending words, so we will output the data set and adjust it manually.
write.csv(Q1SeqTable, "./InputData/Q1Seq.csv", row.names = F)
```

```{r}
# Input the edited data
Q1Sequence <- read.csv("./InputData/Q1Seq_edited.csv")

# Since there are ten IDs from SARS2 don't have search result from NCBI, We decided to drop those ten lines
ID260 <- IDs[!(IDs$virus == "SARS2"), ]
```

### Multiple Alignments
```{r}
Q1DF <- data.frame(ID = ID260$ID,
                   Virus = ID260$virus, 
                   Seq = Q1Sequence$Sequence, 
                   stringsAsFactors = FALSE)

Q1DF <- Q1DF %>% 
  mutate(Transmission_Type = recode(Virus, 
                                    "BCoV1" = "aerosolic", 
                                    "CHIKV" = "vector", 
                                    "DENV" = "vector", 
                                    "Ebola" = "body_fluids", 
                                    "EVA" = "fecal-oral", 
                                    "EVB" = "fecal-oral", 
                                    "EVC" = "fecal-oral",
                                    "EVCr" = "fecal-oral",
                                    "EVD" = "fecal-oral", 
                                    "H3N2" = "aerosolic", 
                                    "HCV" = "blood/sexual",
                                    "HCVr" = "blood/sexual",
                                    "HDV" = "blood/sexual",
                                    "HMPV" = "aerosolic", 
                                    "HRSV" = "aerosolic", 
                                    "HRV3" = "aerosolic", 
                                    "MERS" = "aerosolic",
                                    "MMV" = "aerosolic", 
                                    "MRV" = "aerosolic",
                                    "Norwalk" = "fecal-oral",
                                    "OHVA" = "fecal-oral", 
                                    "PeVA" = "aerosolic",
                                    "RVA" = "aerosolic",
                                    "SARS2" = "aerosolic",   
                                    "SV" = "fecal-oral",
                                    "TBEV" = "vector",
                                    "WNV" = "vector", 
                                    "YFV" = "vector",
                                    "ZIKV" = "vector")) 

# Convert DNASbin to DNAStringSet
VirString <- Q1DF$Seq %>% 
  as.character %>% 
  lapply(., paste0, collapse = "") %>% 
  unlist %>% 
  DNAStringSet 

# Give each sequence a unique names
names(VirString) <- paste(1:nrow(Q1DF), Q1DF$ID, sep = "_")

# Use MUSCLE to align the sequences
VirAlign <- muscle::muscle(stringset = VirString, quiet = T)

# Convert our DNA Multiple Alignment object to a DNA Bin object
VirAlignBin <- as.DNAbin(VirAlign)
```


```{r}
SeqLen <- as.numeric(lapply(VirString, length))
# Show the distribution of sequence length
qplot(SeqLen) + theme_bw()
```

#### **Fig.3** Distribution of sequence length of the selected IDs.


### Visualize the distance matrix
```{r fig.height=15, fig.width=20}
VirDM <- dist.dna(VirAlignBin, model = "K80")
VirDMmat <- as.matrix(VirDM)

# Plot the distance matrix
VirPDat <- melt(VirDMmat)
ggplot(data = VirPDat, aes(x = Var1, y = Var2, fill = value)) + 
  geom_tile() + 
  scale_fill_gradientn(colours = c("white", "blue", "green", "red")) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

#### **Fig.4** Figure of the distance matrix.


### Create the phlogeny tree
```{r}
VirTree <- nj(VirDM)

# Edit the tip label of the tree to better group the sequence by transmission types
VirTree$tip.label <- paste(rownames(Q1DF), Q1DF$Transmission_Type)
TransType <- split(VirTree$tip.label, Q1DF$Transmission_Type)
TransTree <- groupOTU(VirTree, TransType)
```

```{r fig.height=8, fig.width=10}
ggtree(TransTree, branch.length = 'none', layout = "circular", aes(colour = group)) + 
  geom_tiplab(size = 2, aes(angle = angle)) + 
  theme_bw()
```

#### **Fig.5** Phylogenetic tree of the 130 selected sequences without consider the branch length.


### Output the phylogeny tree
```{r}
write.tree(TransTree, "./Output/Transmission_Type_Tree.tre")
```

