---
title: "Biol 432 Group 1 Project"
author: "Chenyang Wu, Caroline Tang"
date: "2022/3/21"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Project Info
#### **Group name**: Teambits
#### **Date**: 2022/3/21
#### **GitHub Link**: https://github.com/carolinetang77/BIOL432-group1

---

### Load the packages we will need
```{r message=FALSE, warning=FALSE}
library(dplyr) 
library(ggplot2)
library(BiocManager)
library(genbankr)
library(rentrez)
library(muscle)
library(ape)
library(reshape2)
library(ggtree)
library(tidyverse)
library(Biostrings)
library(annotate)
```

### Input the data
```{r}
Table1 <- read.csv("./InputData/TableS1.csv")
Table9 <- read.csv("./InputData/TableS9.csv")

readLines("./InputData/TableS8.dat", n = 10) # Take a look at the .dat file
Table8 <- read.table("./InputData/TableS8_edited.dat", header = F) 
names(Table8) <- c("virus", "rate", "year")

Table11 <- read.csv("./InputData/TableS11.csv")
```

---

## **Research question 1**: Are mutation rates correlated with transmission method? Do certain transmission methods have higher mutation rates?

#### Coding by Chenyang Wu

```{r}
# Check how many types of virus in table S9
table(Table9$virus)

# Check the corresponding transmission type to the abbreviation in table S1
list(Table1$Abbreviation)
list(Table1$Transmission)
```

### Combine the relevant columns into 1 dataframe
```{r}
# Label the transmission type to the dataset
dNdSData <- Table9 %>% 
  mutate(Transmission_Type = recode(virus, 
                                    "BCoV1" = "aerosolic", 
                                    "CHIKV" = "vector", 
                                    "DENV" = "vector", 
                                    "Ebola" = "body_fluids", 
                                    "EVA" = "fecal-oral", 
                                    "EVB" = "fecal-oral", 
                                    "EVC" = "fecal-oral",
                                    "EVCr" = "fecal-oral",
                                    "EVD" = "fecal-oral", 
                                    "H3N2" = "aerosolic", 
                                    "HCV" = "blood/sexual",
                                    "HCVr" = "blood/sexual",
                                    "HDV" = "blood/sexual",
                                    "HMPV" = "aerosolic", 
                                    "HRSV" = "aerosolic", 
                                    "HRV3" = "aerosolic", 
                                    "MERS" = "aerosolic",
                                    "MMV" = "aerosolic", 
                                    "MRV" = "aerosolic",
                                    "Norwalk" = "fecal-oral",
                                    "OHVA" = "fecal-oral", 
                                    "PeVA" = "aerosolic",
                                    "RVA" = "aerosolic",
                                    "SARS2" = "aerosolic",   
                                    "SV" = "fecal-oral",
                                    "TBEV" = "vector",
                                    "WNV" = "vector", 
                                    "YFV" = "vector",
                                    "ZIKV" = "vector")) 

# Draw a box plot for mean dNdS data with different transmission types
ggplot(dNdSData, aes(x = Transmission_Type, y = meandNdS, 
                     na.rm = TRUE, fill = Transmission_Type)) + 
  geom_boxplot(alpha = 0.8) +
  theme_bw() + 
  scale_fill_brewer(palette = "Set2") + 
  labs(x = "Transmission Types", y = "Mean dNdS")
```

#### **Fig.1** Boxplot of mean dNdS values among different transmission types.


```{r}
MutaRate <- Table8 %>% 
  mutate(Transmission_Type = recode(virus, 
                                    "BCoV1" = "aerosolic", 
                                    "CHIKV" = "vector", 
                                    "DENV" = "vector", 
                                    "Ebola" = "body_fluids", 
                                    "EVA" = "fecal-oral", 
                                    "EVB" = "fecal-oral", 
                                    "EVC" = "fecal-oral",
                                    "EVCr" = "fecal-oral",
                                    "EVD" = "fecal-oral", 
                                    "H3N2" = "aerosolic", 
                                    "HCV" = "blood/sexual",
                                    "HCVr" = "blood/sexual",
                                    "HDV" = "blood/sexual",
                                    "HMPV" = "aerosolic", 
                                    "HRSV" = "aerosolic", 
                                    "HRV3" = "aerosolic", 
                                    "MERS" = "aerosolic",
                                    "MMV" = "aerosolic", 
                                    "MRV" = "aerosolic",
                                    "Norwalk" = "fecal-oral",
                                    "OHVA" = "fecal-oral", 
                                    "PeVA" = "aerosolic",
                                    "RVA" = "aerosolic",
                                    "SARS2" = "aerosolic",   
                                    "SV" = "fecal-oral",
                                    "TBEV" = "vector",
                                    "WNV" = "vector", 
                                    "YFV" = "vector",
                                    "ZIKV" = "vector")) 

# Draw a box plot for the mutation rate under different transmission types
ggplot(MutaRate, aes(x = Transmission_Type, 
                     y = rate, na.rm = TRUE, 
                     fill = Transmission_Type)) + 
  geom_boxplot(alpha = 0.8) +
  theme_bw() + 
  scale_fill_brewer(palette = "Set2") + 
  labs(x = "Transmission Types", y = "Mutation Rate")
```

#### **Fig.2** Boxplot of mutation rate among different Transmission Types 


### Adjust the sequence ID for SARS2
#### Covid sequences have different IDs from other viruses, not based on genbank accession number so we have to fix that
#### This chunk was done with the help of **Caroline Tang**
```{r}
# Load the virus table and the COVID sequence acknowledgement table
Table3 <- read.delim("InputData/TableS3_GISAID_acknowledgements.dat")

# Merge the COVID rows in the virus table with the acknowledgement table
DataMerged <- merge(Table11[Table11$virus == "SARS2",], Table3, by.x = "ID", by.y = "internID")

# Filter for only rows with GenBank accession IDs
Datafilter <- DataMerged %>% 
  filter(genbank_accession != "?") 

# Replace the alldb ID with the accession ID
Datafilter$ID <- Datafilter$genbank_accession

# Select only the columns from the original virus table and rename them
Datafilter <- Datafilter %>% 
  dplyr::select(1:12)
names(Datafilter) <- names(Table11)

# Replace the original covid rows with the ones we just made
finalIDs <- bind_rows(Table11[Table11$virus != "SARS2",], Datafilter)
```


```{r}
# Check the number of IDs in table S11 for each virus
table(finalIDs$virus)
# Since there are so many IDs, We decided to pick 5 IDs from each virus for question 1

# Random pick 5 IDs for each virus
set.seed(1) # Set seed to make sure the output are constant
IDs <- finalIDs %>%
  group_by(virus) %>%
  sample_n(5)
```

```{r}
# Double check if the code correctly pick 5 IDs from each virus
table(IDs$virus)

# Create the id list
ncbi_ids <- IDs$ID 
```


```{r}
# Search the sequence info from NCBI
Q1Vir <- entrez_fetch(db = "nuccore", id = ncbi_ids, rettype = "fasta") 
Q1Seq <- strsplit(Q1Vir, split = "\n\n", fixed = T)
Q1Seq <- unlist(Q1Seq)

# Use regular expression to edit the search result
header <- gsub("(^>.*genome|*cds|*sequence|*SEQUENCES|*RNA)\\n[ATCG].*", "\\1", Q1Seq)

seq <- gsub("^>.*genome\\n([ATCG].*)", "\\1", Q1Seq)
seq <- gsub("^>.*cds\\n([ATCG].*)", "\\1", seq)
seq <- gsub("^>.*sequence\\n([ATCG].*)", "\\1", seq)
seq <- gsub("^>.*SEQUENCES\\n([ATCG].*)", "\\1", seq)
seq <- gsub("^>.*RNA\\n([ATCG].*)", "\\1", seq)

Q1SeqTable <- data.frame(Name = header, Sequence = seq)
Q1SeqTable$Sequence <- gsub("\n", "", Q1SeqTable$Sequence)

# There are several lines have different ending words, so we will output the data set and adjust it manually.
write.csv(Q1SeqTable, "./InputData/Q1Seq.csv", row.names = F)
```

```{r}
# Input the edited data
Q1Sequence <- read.csv("./InputData/Q1Seq_edited.csv")
```

### Multiple Alignments
```{r}
Q1DF <- data.frame(ID = IDs$ID,
                   Virus = IDs$virus, 
                   Seq = Q1Sequence$Sequence, 
                   stringsAsFactors = FALSE)

Q1DF <- Q1DF %>% 
  mutate(Transmission_Type = recode(Virus, 
                                    "BCoV1" = "aerosolic", 
                                    "CHIKV" = "vector", 
                                    "DENV" = "vector", 
                                    "Ebola" = "body_fluids", 
                                    "EVA" = "fecal-oral", 
                                    "EVB" = "fecal-oral", 
                                    "EVC" = "fecal-oral",
                                    "EVCr" = "fecal-oral",
                                    "EVD" = "fecal-oral", 
                                    "H3N2" = "aerosolic", 
                                    "HCV" = "blood/sexual",
                                    "HCVr" = "blood/sexual",
                                    "HDV" = "blood/sexual",
                                    "HMPV" = "aerosolic", 
                                    "HRSV" = "aerosolic", 
                                    "HRV3" = "aerosolic", 
                                    "MERS" = "aerosolic",
                                    "MMV" = "aerosolic", 
                                    "MRV" = "aerosolic",
                                    "Norwalk" = "fecal-oral",
                                    "OHVA" = "fecal-oral", 
                                    "PeVA" = "aerosolic",
                                    "RVA" = "aerosolic",
                                    "SARS2" = "aerosolic",   
                                    "SV" = "fecal-oral",
                                    "TBEV" = "vector",
                                    "WNV" = "vector", 
                                    "YFV" = "vector",
                                    "ZIKV" = "vector")) 

# Convert DNASbin to DNAStringSet
VirString <- Q1DF$Seq %>% 
  as.character %>% 
  lapply(., paste0, collapse = "") %>% 
  unlist %>% 
  DNAStringSet 

# Give each sequence a unique names
names(VirString) <- paste(1:nrow(Q1DF), Q1DF$ID, sep = "_")

# Use MUSCLE to align the sequences
# This line will take more than 1 hour for R to run it.
VirAlign <- muscle::muscle(stringset = VirString, diags = T, gapopen = -10)

# Convert our DNA Multiple Alignment object to a DNA Bin object
VirAlignBin <- as.DNAbin(VirAlign)
```


```{r}
SeqLen <- as.numeric(lapply(VirString, length))
# Show the distribution of sequence length
qplot(SeqLen) + theme_bw()
```

#### **Fig.3** Distribution of sequence length of the selected IDs.


### Visualize the distance matrix
```{r fig.height=15, fig.width=20}
VirDM <- dist.dna(VirAlignBin, model = "K80")
VirDMmat <- as.matrix(VirDM)

# Plot the distance matrix
VirPDat <- melt(VirDMmat)
ggplot(data = VirPDat, aes(x = Var1, y = Var2, fill = value)) + 
  geom_tile() + 
  scale_fill_gradientn(colours = c("white", "blue", "green", "red")) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

#### **Fig.4** Figure of the distance matrix.


### Create the phlogeny tree
```{r}
VirTree <- nj(VirDM)

# Edit the tip label of the tree to better group the sequence by transmission types
VirTree$tip.label <- paste(rownames(Q1DF), Q1DF$Transmission_Type)
TransType <- split(VirTree$tip.label, Q1DF$Transmission_Type)
TransTree <- groupOTU(VirTree, TransType)
```

```{r fig.height=8, fig.width=10}
ggtree(TransTree, branch.length = 'none', layout = "circular", aes(colour = group)) + 
  geom_tiplab(size = 2, aes(angle = angle)) + 
  theme_bw()
```

#### **Fig.5** Phylogenetic tree of the 135 selected sequences without consider the branch length.


### Output the phylogeny tree
```{r}
write.tree(TransTree, "./Output/Transmission_Type_Tree.tre")
```

### Based on the boxplots, it appears that fecal-oral viruses have the highest mutation rates relative to other methods of transmission. However, they also have the lowest dN/dS rates, suggesting a trade-off between mutation rates and rates of non-synonymous mutations. Due to differences in viral genomes, when creating the tree, there were no common sequences found. As a result, the tree showed all sequences as equally distant from one another.

---
## **Research question 2**: Is the mutation rate correlated with guanine-cytosine content?

#### Coding By Caroline Tang

### Subset sequences (10 per virus)
```{r}
set.seed(1)
virusSubset <- finalIDs %>% 
  group_by(virus) %>% 
  slice_sample(n = 10)
```

### Get sequences from GenBank
```{r}
virusID <- GBAccession(virusSubset$ID)
virusGBK <- read.GenBank(virusID, as.character = TRUE)
```

### Calculate GC content of each sequence and merge with virus types
```{r}
gc <- vector(length = length(virusGBK))
for (i in 1:length(virusGBK)) {
  gc[i] <- length(grep("[gc]", unlist(virusGBK[[i]]))) / length(virusGBK[[i]])
}
gcContent <- data.frame(ID = names(virusGBK), gc = gc)
virusSubset <- merge(virusSubset, gcContent, by = "ID")
```

### Scatterplot of GC content vs. mutation rate and transmission method
```{r}
# Calculate mean mutation rate per virus and merge with GC content
mutationMean <- Table8 %>% 
  group_by(virus) %>% 
  summarise(meanRate = mean(rate, na.rm = T))
virusSubset <- merge(virusSubset, mutationMean, by = "virus")

# Merge transmission data
virusSubset <- merge(virusSubset, Table1, by.x = "virus", by.y = "Abbreviation")

# Create scatter plot
ggplot(data = virusSubset, aes(x = gc, y = meanRate)) +
  geom_point(aes(colour = Transmission), alpha = 0.8) +
  theme_classic() +
  geom_smooth(method = "lm") +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "GC content", y = "Mean mutation rate")
```

#### **Fig.6** Scatter plot of the 130 selected sequences' GC content and mutation rates, colour coded by transmission type.  

### Based on the scatter plot, there is a slightly positive correlation between GC content and mutation rate, which supports the hypothesis that mutation rate increases with GC content. However, this trend varied among transmission methods, and the overall trend be skewed by the fecal-oral viruses with high mutation rates.
